src/sumo_store_riak.erl

```diff
@@ -61,23 +68,38 @@
 -spec init(
   term()
 ) -> {ok, term()}.
-init(Options) ->
+init(Opts) ->
   % The storage backend key in the options specifies the name of the process
   % which creates and initializes the storage backend.
-  Backend = proplists:get_value(storage_backend, Options),
-  State = sumo_backend_riak:get_state(Backend),
+  Backend = proplists:get_value(storage_backend, Opts),
+  Conn = sumo_backend_riak:get_connection(Backend),
+  BucketType = iolist_to_binary(proplists:get_value(bucket_type, Opts)),
+  Bucket = iolist_to_binary(proplists:get_value(bucket, Opts, <<"sumo">>)),
+  Index = iolist_to_binary(proplists:get_value(index, Opts, <<"sumo_idx">>)),
+  R = proplists:get_value(r_args, Opts, []),
+  W = proplists:get_value(w_args, Opts, []),
+  State = #state{backend = Backend, conn = Conn,
+                 bucket = {BucketType, Bucket},
+                 index = Index,
+                 r_args = R,
+                 w_args = W},
   {ok, State}.
 
 -spec persist(
   sumo_internal:doc(), state()
 ) -> sumo_store:result(sumo_internal:doc(), state()).
-persist(Doc, #state{conn = Conn, bucket = Bucket} = State) ->
-  {Id, NewDoc} = new_doc(Doc, State),
-  case update_map(Conn, Bucket, Id, doc_to_rmap(NewDoc)) of
-    {error, Error} ->
-      {error, Error, State};
-    _ ->
-      {ok, NewDoc, State}
+persist(Doc, #state{backend = Backend, bucket = Bucket, w_args = W} = State) ->
+  Conn = sumo_backend_riak:checkout_conn(Backend),
+  try
+    {Id, NewDoc} = new_doc(Doc, Conn, State),
+    case update_map(Conn, Bucket, Id, doc_to_rmap(NewDoc), W) of
+      {error, Error} ->
+        {error, Error, State};
+      _ ->
+        {ok, NewDoc, State}
+    end
+  after
+    ok = sumo_backend_riak:checkin_conn(Backend, Conn)
```

Comment:
@elbrujohalcon is the return value of this always ok? I don't know how after works as a last statement.